lib/metrics.js

```diff
@@ -0,0 +1,848 @@
+    this.reset(this.duration, mixMeta(this.#meta, meta));
+    return this.duration;
+  }
+
+  /**
+   * Support `using` syntax to automatically stop the timer when done.
+   */
+  [SymbolDispose]() {
+    this.stop();
+  }
+}
+
+/**
+ * A count of the number of unique values that have been seen.
+ */
+class UniqueSet extends Gauge {
+  // Use a weak set to track unique values without retaining references.
+  #seen = new SafeWeakSet();
+
+  /**
+   * The number of unique values seen.
+   * @property {number} count
+   */
+  count = 0;
+
+  /**
+   * Add value to set. If value was not already present, report it.
+   * @param {any} value The value to track in the set.
+   * @param {object} meta Additional metadata to include with the report.
+   */
+  add(value, meta) {
+    // If already seen, do nothing.
+    if (this.#seen.has(value)) {
+      return;
+    }
+
+    this.count += 1;
+    this.#seen.add(value);
+    this.reset(this.count, meta);
+  }
+}
+
+/**
+ * A gauge which periodically updates its value by calling a function and
+ * setting the value to the result.
+ */
+class PeriodicGauge extends Gauge {
+  #timer;
+  #interval;
+  #fn;
+
+  /**
+   * Construct a new periodic gauge.
+   * @param {Metric} metric The metric to report to.
+   * @param {number} interval The interval in milliseconds to update the gauge.
+   * @param {Function} fn The function to call to update the gauge.
+   */
+  constructor(metric, interval, fn) {
+    super(metric);
+
+    if (typeof interval !== 'number' || interval <= 0) {
+      throw new ERR_INVALID_ARG_TYPE('interval', ['number'], interval);
+    }
+    if (typeof fn !== 'function') {
+      throw new ERR_INVALID_ARG_TYPE('fn', ['function'], fn);
+    }
+
+    this.#timer = undefined;
+    this.#interval = interval;
+    this.#fn = fn;
+
+    this.schedule();
+  }
+
+  /**
+   * Schedule the update timer.
+   */
+  schedule() {
+    this.stop();
+
+    this.#timer = setInterval(() => {
+      this.reset(this.#fn());
+    }, this.interval);
+
+    // Don't keep the process alive just for this timer.
+    this.#timer.unref();
+  }
+
+  /**
+   * The interval in milliseconds at which to update the value. If changed,
+   * the timer will be rescheduled.
+   * @property {number} interval
+   */
+  set interval(interval) {
+    if (typeof interval !== 'number' || interval <= 0) {
+      throw new ERR_INVALID_ARG_TYPE('interval', ['number'], interval);
+    }
+    this.#interval = interval;
+    this.schedule();
+  }
+  get interval() {
+    return this.#interval;
+  }
+
+  /**
+   * Stop the periodic gauge.
+   */
+  stop() {
+    if (this.#timer !== undefined) {
+      clearInterval(this.#timer);
+      this.#timer = undefined;
+    }
+  }
+
+  /**
+   * Reference the timer to prevent to loop from exiting.
+   */
+  ref() {
+    this.#timer?.ref();
+  }
+
+  /**
+   * Unreference the timer to allow the loop to exit.
+   */
+  unref() {
+    this.#timer?.unref();
+  }
```

Comment:
Would prefer to avoid adding new APIs with `ref()`/`unref()`... these should use `process.ref/unref` (https://nodejs.org/docs/latest/api/process.html#processrefmayberefable)

Reply:
Ah, I entirely missed that getting added. Definitely makes sense, especially in this case with timers. Never like that timers didn't quite follow the spec because of the need for ref/unref functions. 😅